### ISSUE TITLE
Fail to consider a member access from an object to a static member a constant expression

### ISSUE THREAD

**Issue by Sainan** [llvm/llvm-project#57254]
```C++
#include <iostream>

struct Usertype
{
    [[nodiscard]] static constexpr bool getSomething() noexcept
    {
        return true;
    }
};

void checkSomething(Usertype& ud)
{
    if constexpr (ud.getSomething()) // on linux, only works if ud is not a reference or r-value
    //if constexpr (std::remove_reference_t<decltype(ud)>::getSomething()) // always works
    {
        std::cout << "Alright!\n";
    }
}

int main()
{
    Usertype ud;
    checkSomething(ud);
}
```

This compiles fine on Windows, but on Linux I get "constexpr if condition is not a constant expression".

Tested against clang versions:
- Windows: 14.0.5, 13.0.0
- Linux: 14.0.0, 13.0.1, 13.0.0, 10.0.0-4ubuntu1, 7.0.1-8+deb10u2

Using C++17 in all instances.

**Comment by tbaederr** [comment]
Using `Usertype::getSomething()` works.

**Comment by Sainan** [comment]
Right, I was looking at this from the perspective of the type being hidden due to a macro that defines a template function. In any case, I have workarounds, it's not a show-stopper of a problem, but just really intriguing that the build only fails on Linux.

**Comment by tbaederr** [comment]
> the type being hidden due to a macro that defines a template function. 

Wait, where's a macro involved? And where a template function?

I was thinking it might be due to a different stdlib, but

```c++
struct Usertype
{
    static constexpr bool getSomething()
    {
        return true;
    }
};

void checkSomething(Usertype& ud)
{
    if constexpr (ud.getSomething()) {}
}
```

Also reproduces.

**Comment by Sainan** [comment]
> Wait, where's a macro involved? And where a template function?

Not in the example, but just explaining how I was looking at it that I didn't think of `Usertype::getSomething()`.

**Comment by AaronBallman** [comment]
This looks very closely related to #56246 -- that one is about resolving to a static data member of `this`, but this one is through a member object expression other than `this`.

**Comment by usx95** [comment]
```cpp
struct Foo {
    static constexpr bool Bar() { return true; }
};

void func(Foo& foo) { 
    static_assert(foo.Bar());
}
```
`foo.Bar()` is not a constant expression because `foo`'s lifetime began outside the expression and it is not initialised with a constant expression.
http://eel.is/c++draft/expr.const#4

Related discussions: https://github.com/llvm/llvm-project/issues/40438
Related Patch: https://github.com/llvm/llvm-project/commit/00068c452a599c328986e8afcbb3311331d09d26#diff-24549b124046a492279fe53ffe2aca86550f281b025a48067c043a1c944547faR2306-R2323

Can you confirm @AaronBallman

**Comment by AaronBallman** [comment]
> Can you confirm @AaronBallman

I think this is the subject of https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2022/p2280r4.html which we don't claim to implement yet. (Thanks @shafik for noticing that connection!)

**Comment by cor3ntin** [comment]
Should be fixed by #63139

**Comment by zygoloid** [comment]
Closing as duplicate of #63139.
